lib/config.js

```diff
@@ -212,127 +87,276 @@ class Config {
+                continue;
+            }
 
-        // Only override parser if it is passed explicitly through the command line or if it's not
-        // defined yet (because the final object will at least have the parser key)
-        if (this.parser || !config.parser) {
-            config = ConfigOps.merge(config, {
-                parser: this.parser
-            });
-        }
+            // Check for root flag
+            if (localConfig.root === true) {
+                rootPath = path.dirname(localConfigFile);
+            }
 
-        if (this.parserOptions) {
-            config = ConfigOps.merge(config, {
-                parserOptions: this.parserOptions
-            });
+            debug(`Using ${localConfigFile}`);
+            configs.push(localConfig);
+            searched.push(localConfigDirectory);
         }
 
-        // Step 8: Merge in command line plugins
-        if (this.options.plugins) {
-            debug("Merging command line plugins");
-            Plugins.loadAll(this.options.plugins);
-            config = ConfigOps.merge(config, { plugins: this.options.plugins });
-        }
+        if (!configs.length && !cache && !this.useSpecificConfig) {
 
-        // Step 9: Apply environments to the config if present
-        if (config.env) {
-            config = ConfigOps.applyEnvironments(config);
+            // Fall back on the personal config from ~/.eslintrc
+            debug("Using personal config file");
+            const personalConfig = this.getPersonalConfig();
+
+            if (personalConfig) {
+                configs.push(personalConfig);
+            } else if (!hasRules(this.options) && !this.options.baseConfig) {
+
+                // No config file, no manual configuration, and no rules, so error.
+                const noConfigError = new Error("No ESLint configuration found.");
+
+                noConfigError.messageTemplate = "no-config-found";
+                noConfigError.messageData = {
+                    directory,
+                    filesExamined: localConfigFiles
+                };
+
+                throw noConfigError;
+            }
         }
 
-        this.cache[directory] = config;
+        // Merged with any cached portion
+        configs.reverse();
+        cache = cache ? cache.concat(configs) : configs;
 
-        return config;
+        // Set the caches for the parent directories
+        configCache.setHierarchyLocalConfigs(searched, cache);
+
+        return cache;
+    }
+
+    /**
+     * Gets the vector of applicable configs from the hierarchy for a given file, including any overrides that apply to
+     * the specified file path. A vector is an array of config file paths, each optionally followed by one or more
+     * numbers that correspond to the indices of nested config blocks within the config file's overrides section whose
+     * glob patterns match the specified file path; e.g., the vector ['/home/john/project/app/.eslintrc', 0, 2] would
```

Comment:
Is there a reason to use an array with strings and numbers here, rather than some other data structure? An array containing multiple datatypes with different meanings seems like a code smell.

Maybe an object would be better:

```js
{
  configFilePath: '/home/john/project/app/.eslintrc',
  overrideConfigs: [
    {
      rules: { /* ... */ }
    },
    {
      rules: { /* ... */ }
    }
  ]
}
```